{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "mighty-mobile",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-08T04:37:57.809348Z",
     "start_time": "2022-09-08T04:37:55.556896Z"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "# import scipy.stats as ss\n",
    "# from scipy.spatial import distance_matrix\n",
    "import random\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from sklearn.cluster import SpectralClustering\n",
    "from sklearn.metrics.cluster import adjusted_rand_score\n",
    "from tqdm.notebook import tqdm\n",
    "from collections import defaultdict\n",
    "from multiprocessing import Pool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "spoken-stylus",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-08T04:37:57.814657Z",
     "start_time": "2022-09-08T04:37:57.811898Z"
    }
   },
   "outputs": [],
   "source": [
    "INPUT_PATH=\"/mnt/geofflab/SNP_barcoding/Lvar_annotations_v3_Jan2021/transcripts.csv\"\n",
    "COUNTS_PATH=\"/mnt/geofflab/SNP_barcoding/LV_counts_10hpf.csv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dietary-minutes",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-08T04:37:57.826941Z",
     "start_time": "2022-09-08T04:37:57.816940Z"
    }
   },
   "outputs": [],
   "source": [
    "heterozygosity = 0.001 # This is the rate of SNPs per base\n",
    "alphabet = {\"A\":0, \"C\":1, \"G\":2, \"T\":3}\n",
    "\n",
    "# genotypes where chromosome order matters (used internally during generation)\n",
    "genotypes_internal = {\n",
    "    \"AA\":0, \"AC\":1, \"AG\":2, \"AT\":3, \n",
    "    \"CA\":4, \"CC\":5, \"CG\":6, \"CT\":7,\n",
    "    \"GA\":8, \"GC\":9, \"GG\":10, \"GT\":11, \n",
    "    \"TA\":12, \"TC\":13, \"TG\":14, \"TT\":15,\n",
    "}\n",
    "\n",
    "# genotypes where chromosome order doesn't matter (what we see)\n",
    "genotypes_external = {\n",
    "\"AA\":0, \"AC\":1, \"AG\":2, \"AT\":3, \n",
    "\"CA\":1, \"CC\":4, \"CG\":5, \"CT\":6,\n",
    "\"GA\":2, \"GC\":4, \"GG\":7, \"GT\":8, \n",
    "\"TA\":3, \"TC\":6, \"TG\":8, \"TT\":9,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "revolutionary-sally",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-08T04:42:52.952246Z",
     "start_time": "2022-09-08T04:42:52.917474Z"
    }
   },
   "outputs": [],
   "source": [
    "def sample_dropout_profile(args):\n",
    "    \"\"\"\n",
    "    Sample a dropout profile from single-cell transcript read count df, and \n",
    "    introduce dropout over the snp_profile according to the dropout profile.\n",
    "    \"\"\"\n",
    "    global count_df\n",
    "    snps = args[0].copy()\n",
    "    loci_map = args[1]\n",
    "    \n",
    "    np.random.seed()\n",
    "    \n",
    "    # Sample a cell's counts\n",
    "    sampled_dprofile = np.random.choice(count_df.columns.tolist()[2:])\n",
    "\n",
    "    for idx,row in count_df[[\"Lv_name\",sampled_dprofile]].iterrows():\n",
    "        gene,count = row\n",
    "\n",
    "        # If we have no counts, then we have can't detect the genotype\n",
    "        if count <= 1:\n",
    "            for locus in loci_map[gene]:\n",
    "                snps[locus] = -1 # -1 genotype -> undetected\n",
    "        else:\n",
    "            # Otherwise compute the probability of only seeing one of the chromosomes, or both\n",
    "            prob_one = 0.5**count\n",
    "            prob_both = 1 - 2*prob_one\n",
    "\n",
    "            # Now choose whether we only see the first chromosome, second chromosome, or both\n",
    "            seen = np.random.choice([0, 1, 2], p=[prob_one, prob_one, prob_both])\n",
    "\n",
    "            if seen == 0:\n",
    "                # Convert internal to external genotypes on the first chromosome\n",
    "                first_mapping = {0:0, 1:0, 2:0, 3:0,\n",
    "                                 4:4, 5:4, 6:4, 7:4,\n",
    "                                 8:7, 9:7, 10:7, 11:7,\n",
    "                                 12:9, 13:9, 14:9, 15:9}\n",
    "\n",
    "                # Convert the SNP read to it's value on one of the chromosomes\n",
    "                for snp_loc in loci_map[gene]:\n",
    "                    snps[snp_loc] = first_mapping[snps[snp_loc]]\n",
    "\n",
    "            elif seen == 1:\n",
    "                # Convert internal to external genotypes on the first chromosome\n",
    "                second_mapping = {0:0, 1:4, 2:7, 3:9,\n",
    "                                 4:0, 5:4, 6:7, 7:9,\n",
    "                                 8:0, 9:4, 10:7, 11:9,\n",
    "                                 12:0, 13:4, 14:7, 15:9}\n",
    "\n",
    "                # Convert the SNP read to it's value on one of the chromosomes\n",
    "                for snp_loc in loci_map[gene]:\n",
    "                    snps[snp_loc] = second_mapping[snps[snp_loc]]\n",
    "\n",
    "            else:\n",
    "\n",
    "                # Convert internal to external genotypes on both chromosomes\n",
    "                both_mapping = {0:0, 1:1, 2:2, 3:3,\n",
    "                                 4:1, 5:4, 6:5, 7:6,\n",
    "                                 8:2, 9:4, 10:7, 11:8,\n",
    "                                 12:3, 13:6, 14:8, 15:9}\n",
    "\n",
    "                # Convert the SNP read to it's value on one of the chromosomes\n",
    "                for snp_loc in loci_map[gene]:\n",
    "                    snps[snp_loc] = both_mapping[snps[snp_loc]]\n",
    "\n",
    "    return snps\n",
    "\n",
    "class EmbryoSequenceSimulator:\n",
    "    def __init__(self, input_df, n_embryos):\n",
    "        \n",
    "        # pd data frame\n",
    "        self.input_df = input_df # cols: gene,chromosome,start,end,direction,sequence\n",
    "        self.n_embryos = n_embryos\n",
    "        \n",
    "        # Number of cells that can be harvested from embryo\n",
    "        self.avg_harvested_cells = 20\n",
    "        self.sd_harvested_cells = 3\n",
    "        self.alphabet = {\"A\":0, \"C\":1, \"G\":2, \"T\":3}\n",
    "        \n",
    "        # Global parental chromatid maps\n",
    "        self.m0_chromatid = None\n",
    "        self.m1_chromatid = None\n",
    "        self.p0_chromatid = None\n",
    "        self.p1_chromatid = None\n",
    "        \n",
    "    def global_seeding(self, heterozygosity):\n",
    "        \"\"\"\n",
    "        Uniform seeding of SNP locations for each gene with frequency according to heterozygosity.\n",
    "        Parental chromatid seeding (2 chromatids from mother, 2 chromatids from father)\n",
    "        \"\"\"\n",
    "        \n",
    "        # All the SNPs present in the parents\n",
    "        snp_map_overall = None\n",
    "        \n",
    "        # Setting parental chromatid maps\n",
    "        for parent in [\"m0\",\"m1\",\"p0\",\"p1\"]:\n",
    "            \n",
    "            # Create a profile of SNPs for each parent chromosome\n",
    "            snp_map = defaultdict(dict)\n",
    "            sequence_map = defaultdict(dict)\n",
    "            \n",
    "            for idx,transcript_entry in self.input_df.iterrows():\n",
    "                gene,chrom,start,end,direction,sequence = transcript_entry\n",
    "                N = len(sequence)\n",
    "                num_snps = round(N*heterozygosity)\n",
    "                snp_positions = list(np.sort(random.sample(range(N),num_snps)))\n",
    "                snp_map[chrom][gene] = snp_positions\n",
    "                sequence_map[chrom][gene] = sequence\n",
    "                \n",
    "            # Add the snp map we just made to the global collection of SNPs\n",
    "            if snp_map_overall is None:\n",
    "                snp_map_overall = snp_map\n",
    "            else:\n",
    "                for idx,transcript_entry in self.input_df.iterrows():\n",
    "                    gene,chrom,start,end,direction,sequence = transcript_entry\n",
    "                    snp_map_overall[chrom][gene].extend(snp_map[chrom][gene])\n",
    "            \n",
    "            chromatid = defaultdict(dict)\n",
    "            # If we have a SNP, this is the probability of going reference to \n",
    "            transition_probs = np.array([[0, 1/3, 1/3, 1/3],\n",
    "                                         [1/3, 0, 1/3, 1/3],\n",
    "                                         [1/3, 1/3, 0, 1/3],\n",
    "                                         [1/3, 1/3, 1/3, 0]])\n",
    "            \n",
    "            for chrom,genes in snp_map.items():            \n",
    "                for gene,snp_positions in genes.items():\n",
    "                    sequence_list = [char for char in sequence_map[chrom][gene]]\n",
    "                    snp_profile = {locus:np.random.choice(list(self.alphabet.keys()), \n",
    "                                                          p=transition_probs[alphabet[sequence_list[locus]]]) for locus in snp_positions}\n",
    "                    \n",
    "                    for locus,char in snp_profile.items():\n",
    "                        sequence_list[locus] = char\n",
    "                    \n",
    "                    chromatid[chrom][gene] = \"\".join(sequence_list)\n",
    "                \n",
    "            # Assign seeded parental chromatid to its class field (global over all embryos)\n",
    "            if parent == \"m0\":\n",
    "                self.m0_chromatid = chromatid\n",
    "            elif parent == \"m1\":\n",
    "                self.m1_chromatid = chromatid\n",
    "            elif parent == \"p0\":\n",
    "                self.p0_chromatid = chromatid\n",
    "            elif parent == \"p1\":\n",
    "                self.p1_chromatid = chromatid\n",
    "                \n",
    "        return snp_map_overall\n",
    "    \n",
    "    def embryo_population_simulation(self, count_df=None):\n",
    "        \"\"\"\n",
    "        Simulating SNP profile for each embryo across all transcript entries in input_df.\n",
    "        For this simulator, not explicitly generating scRNA-seq reads for each embryo (reduces space)\n",
    "        \"\"\"\n",
    "\n",
    "        # map: idx of transcript entry -> list of snp positions\n",
    "        snp_map = self.global_seeding(heterozygosity)\n",
    "        true_clusters = []\n",
    "        cluster_label = 0\n",
    "        \n",
    "        population_profile = [] # The variant calls for the cells\n",
    "        \n",
    "        p = Pool(NUM_PROC)\n",
    "        \n",
    "        for embryo in range(self.n_embryos):\n",
    "            \n",
    "            # Each entry a list of genotypes for an embryo across all transcripts\n",
    "            snp_profile,gene_locus_map = self.determine_embryo_genotype(snp_map)\n",
    "\n",
    "            # Harvest some number of cells, which will theoretically have identical SNP profiles\n",
    "            n_harvested_cells = round(np.random.normal(self.avg_harvested_cells,self.sd_harvested_cells))\n",
    "            \n",
    "            # Parallelize making the cells\n",
    "            new_cells = list(tqdm(p.imap(sample_dropout_profile, \n",
    "                                    zip([snp_profile]*n_harvested_cells, \n",
    "                                        [gene_locus_map]*n_harvested_cells)), \n",
    "                                  total=n_harvested_cells, desc=f'Embryo {embryo}'))\n",
    "                \n",
    "            population_profile.extend(new_cells)\n",
    "\n",
    "            \n",
    "            # Mark the cluster of the cells we just created\n",
    "            for _ in range(n_harvested_cells):\n",
    "                true_clusters.append(cluster_label)\n",
    "            \n",
    "            cluster_label += 1\n",
    "            \n",
    "        p.close()\n",
    "                \n",
    "        return np.array(population_profile),true_clusters, gene_locus_map\n",
    "        \n",
    "    \n",
    "    def determine_embryo_genotype(self, snp_map):\n",
    "        \"\"\"\n",
    "        Simulate array containing genotypes over all transcripts \n",
    "        (concatenated into single list)\n",
    "        \"\"\"\n",
    "                \n",
    "        # state (genotype) across all transcripts\n",
    "        snp_genotypes,idx = [],0\n",
    "        gene_locus_map = defaultdict(list)\n",
    "                \n",
    "        # map: idx of transcript entry -> sequence after homologous recombination\n",
    "        m_chromatid,p_chromatid = self.simulate_embryo_sequence()\n",
    "        for chrom,genes in snp_map.items():\n",
    "            for gene,snp_positions in genes.items():                \n",
    "                m_sequence = m_chromatid[chrom][gene]\n",
    "                p_sequence = p_chromatid[chrom][gene]\n",
    "                for locus in snp_positions:\n",
    "                    snp_genotype = \"\".join([m_sequence[locus],p_sequence[locus]])\n",
    "                    snp_genotypes.append(genotypes_internal[snp_genotype])\n",
    "                    gene_locus_map[gene].append(idx)\n",
    "                    idx += 1\n",
    "                \n",
    "        return snp_genotypes,gene_locus_map\n",
    "                \n",
    "    def simulate_embryo_sequence(self):\n",
    "        \"\"\"\n",
    "        Given seeded parental chromatids and a map of SNP locations for each transcript,\n",
    "        simulate embryo sequences\n",
    "        \"\"\"\n",
    "        \n",
    "        updated_m0,updated_m1 = defaultdict(dict),defaultdict(dict)\n",
    "        updated_p0,updated_p1 = defaultdict(dict),defaultdict(dict)\n",
    "        \n",
    "        for parent in [\"mother\",\"father\"]:\n",
    "            \n",
    "            # Consider homologous recombination with respect to each chromosome\n",
    "            for chr_num in np.sort(np.unique(df[\"chromosome\"])):\n",
    "                \n",
    "                # Subset df for specific chr, sorted by start position \n",
    "                chr_df = self.input_df[self.input_df[\"chromosome\"] == chr_num].sort_values(by=[\"start\"])\n",
    "                \n",
    "                # Chiasma point chosen over range of transcripts reads\n",
    "                transcript_range = range(chr_df[\"start\"].iloc[0], chr_df[\"start\"].iloc[-1]+len(chr_df[\"sequence\"].iloc[-1])) \n",
    "                chiasma = random.choice(transcript_range)\n",
    "                chiasma_flag = 1 # 1 means chiasma has not been reached; 0 means passed chiasma\n",
    "                for idx,transcript_entry in chr_df.iterrows():\n",
    "                    if parent == \"mother\":\n",
    "                        chiasma_flag = self.homologous_recombination(\n",
    "                            transcript_entry = transcript_entry,\n",
    "                            updated_c0 = updated_m0,\n",
    "                            updated_c1 = updated_m1,\n",
    "                            c0 = self.m0_chromatid,\n",
    "                            c1 = self.m1_chromatid,\n",
    "                            chiasma = chiasma,\n",
    "                            flag = chiasma_flag\n",
    "                        )\n",
    "                    else:\n",
    "                        chiasma_flag = self.homologous_recombination(\n",
    "                            transcript_entry = transcript_entry,\n",
    "                            updated_c0 = updated_p0,\n",
    "                            updated_c1 = updated_p1,\n",
    "                            c0 = self.p0_chromatid,\n",
    "                            c1 = self.p1_chromatid,\n",
    "                            chiasma = chiasma,\n",
    "                            flag = chiasma_flag\n",
    "                        )\n",
    "\n",
    "        m_chromatid = self.chromosome_inheritance(updated_m0, updated_m1)\n",
    "        p_chromatid = self.chromosome_inheritance(updated_p0, updated_p1)\n",
    "        \n",
    "        return m_chromatid,p_chromatid\n",
    "    \n",
    "    def homologous_recombination(self,transcript_entry,updated_c0,updated_c1,c0,c1,chiasma,flag):\n",
    "        \"\"\"\n",
    "        for a given transcript entry, simulate process of homologous recombination, given\n",
    "        a uniformly selected chiasma point. \n",
    "        \"\"\"\n",
    "        \n",
    "        gene,chrom,start,end,direction,sequence = transcript_entry\n",
    "        \n",
    "        # 1 means chiasma has not been reached; 0 means passed chiasma\n",
    "        if flag <= start and flag == 1:\n",
    "            flag = 0\n",
    "            \n",
    "            # Case where chiasma point lies within a gene\n",
    "            if chiasma in range(start,start+len(sequence)):\n",
    "                updated_c0[chrom][gene] = c0[chrom][gene][:chiasma]+c1[chrom][gene][chiasma:]\n",
    "                updated_c1[chrom][gene] = c1[chrom][gene][:chiasma]+c0[chrom][gene][chiasma:]\n",
    "                return flag\n",
    "\n",
    "        if flag == 1: # Before chiasma point has been reached\n",
    "            updated_c0[chrom][gene] = c0[chrom][gene]\n",
    "            updated_c1[chrom][gene] = c1[chrom][gene]\n",
    "        else: # After chiasma point has been reached\n",
    "            updated_c0[chrom][gene] = c1[chrom][gene]\n",
    "            updated_c1[chrom][gene] = c0[chrom][gene]\n",
    "        \n",
    "        return flag\n",
    "    \n",
    "    def chromosome_inheritance(self,c0,c1):\n",
    "        \"\"\"\n",
    "        Given homologous chromosomes, randomly select one to be inherited by embryo\n",
    "        \"\"\"\n",
    "        \n",
    "        chromatid = defaultdict(dict)\n",
    "        for chrom in c0:\n",
    "            chromatid[chrom] = c0[chrom] if random.random() <= 0.5 else c1[chrom]\n",
    "        \n",
    "        return chromatid\n",
    "                  \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "mediterranean-ireland",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-08T04:42:54.579391Z",
     "start_time": "2022-09-08T04:42:54.575567Z"
    }
   },
   "outputs": [],
   "source": [
    "n_embryos = 10\n",
    "NUM_PROC = 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "according-circular",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-08T04:38:30.150866Z",
     "start_time": "2022-09-08T04:37:57.990281Z"
    }
   },
   "outputs": [],
   "source": [
    "df = pd.read_csv(INPUT_PATH)\n",
    "df = df.dropna()\n",
    "\n",
    "count_df = pd.read_csv(COUNTS_PATH)\n",
    "count_df = count_df[count_df[\"Lv_name\"] != \"LVA_m28240.t1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "distributed-briefing",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-08T04:39:23.866256Z",
     "start_time": "2022-09-08T04:38:30.153123Z"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "760759f4c1c34431a8856a9210614f99",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/29836 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Dowsample all sequences to max 150 bp\n",
    "for ind in tqdm(df.index):\n",
    "    start = df.loc[ind, 'start']\n",
    "    end = df.loc[ind, 'end']\n",
    "    sequence = df.loc[ind, 'sequence']\n",
    "    \n",
    "    # If a sequence is longer than 150, trim it\n",
    "    if np.absolute(end-start) > 150:\n",
    "        df.loc[ind, 'end'] = start + 150\n",
    "        df.loc[ind, 'sequence'] = sequence[:150]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bound-cruise",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-08T04:45:06.751203Z",
     "start_time": "2022-09-08T04:42:59.674521Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "655c24d230964b2082136d4212d6894d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Embryo 0:   0%|          | 0/20 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "137c2382db734edda007eb5d055c961f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Embryo 1:   0%|          | 0/22 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dd82c2153fe7418a8ca82df2a25ff8b2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Embryo 2:   0%|          | 0/23 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ad3607371b94452ab879abfe2cbf89f6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Embryo 3:   0%|          | 0/17 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a1bd9234d8ab4b5e9b6c25cc43fc93aa",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Embryo 4:   0%|          | 0/21 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "93372445cfb2459ab6fab8fb4b458e13",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Embryo 5:   0%|          | 0/22 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "07327c633f434c02bfda0be69e22aff2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Embryo 6:   0%|          | 0/17 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "be3661989a8f432791fba2c33aaeffac",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Embryo 7:   0%|          | 0/21 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5d815f2a3858457f86b7182276592bf1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Embryo 8:   0%|          | 0/20 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3fe441e615214b2fa9aec53ffde04077",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Embryo 9:   0%|          | 0/19 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(202, 0)\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "simulator = EmbryoSequenceSimulator(df, n_embryos)\n",
    "X,true_clusters,gene_locus_map = simulator.embryo_population_simulation(count_df)\n",
    "print(X.shape)\n",
    "print(X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "elder-definition",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-08T04:41:44.861530Z",
     "start_time": "2022-09-08T04:41:43.649685Z"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "97c268b5d6334d3fa2765dfeb765cade",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/207 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "n = X.shape[0]\n",
    "\n",
    "def adjustedHamming(g0, g1):\n",
    "# Computes a hamming distance normalized by where we have\n",
    "# data. ie. entries are not -1.\n",
    "    if len(g0) != len(g1):\n",
    "        raise ValueError(\"Sequence lengths unequal.\")\n",
    "    else:\n",
    "        hd,adjlen = 0, 0\n",
    "        for i, j in zip(g0, g1):\n",
    "            if i == -1 or j == -1: \n",
    "                continue\n",
    "            \n",
    "            adjlen += 1\n",
    "            \n",
    "            if i != j: \n",
    "                hd += 1\n",
    "                \n",
    "        if adjlen == 0:\n",
    "            return np.inf\n",
    "        else:\n",
    "            return hd/adjlen\n",
    "\n",
    "def compute_row_ham_dists(i):\n",
    "# Compute a row of hamming dists\n",
    "    row_hd = np.zeros(n-i)\n",
    "    \n",
    "    for j in range(i, n):\n",
    "        hd = adjustedHamming(X[i,:], X[j,:])\n",
    "        row_hd[j-i] = hd\n",
    "        \n",
    "    return row_hd\n",
    "\n",
    "\n",
    "with Pool(NUM_PROC) as p:\n",
    "    ham_dists = list(tqdm(p.imap(compute_row_ham_dists, range(n)), total=n))\n",
    "\n",
    "hd = np.zeros((n, n))\n",
    "\n",
    "for i in range(n):\n",
    "    hd[i, i:] = ham_dists[i]\n",
    "\n",
    "hd = hd + hd.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "significant-portsmouth",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-08T04:41:45.123045Z",
     "start_time": "2022-09-08T04:41:44.864007Z"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhIAAAGfCAYAAAAK+zskAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAk7ElEQVR4nO3df6zldX3n8eero7hp1QXKgCM/FnSnbLHZRTpBE1e3XaQC6Yp2Y3foxlIlOyWBrG7bpGNNuqbGhNqqiamFjJWIGwVp0TpppkUkpqaJKANFZEDKQFEGpsMAXTFLC52Z9/5xvlcP95x772e+59x7z708H+Sbc76f7/d7zud7zr3D+77fn+/nm6pCkiSpjx9b7Q5IkqS1y0BCkiT1ZiAhSZJ6M5CQJEm9GUhIkqTeDCQkSVJvBhKSJM2gJBckuT/J3iTbx2z/d0m+nuTZJL/VcmyS45PckuSB7vG4Sfu5bIHEUh+AJEkaL8kG4BPAhcBZwCVJzpq321PA/wT+8CiO3Q7cWlWbgVu79YksSyDR+AFIkqTxzgX2VtVDVfUccANw8fAOVfV4Vd0O/MtRHHsxcF33/DrgbZN29EWTvsACfngSAEnmTuLecTufcMIJdfrppy9TVyRJ690dd9zxRFVtXM73eMvP/0Q9+dThqbzWHXc/uwf456GmHVW1Y2j9ZOCRofV9wOsaX36xY0+qqv0AVbU/yYlH1fExliuQWPIDSLIN2AZw2mmnsXv37mXqiiRpvUvy3eV+jyefOsw3bz5tKq+1YdMD/1xVWxbZJWPaWu9pMcmxR225xkgseRJVtaOqtlTVlo0blzWIlCRpYgUcmdJ/DfYBpw6tnwI81tjVxY49kGQTQPf4eONrLmi5AolJPgBJkl7obgc2JzkjyTHAVmDnFI7dCVzaPb8U+NKkHV2u0sYPTwJ4lMFJ/MoyvZckSSugOFxN2YTJ36nqUJIrgZuBDcC1VbUnyeXd9muSvALYDbwcOJLkvcBZVfX0uGO7l74KuDHJZcD3gHdM2tdlCSQW+gCW470kSVoJg9LGsg01GH2/ql3Arnlt1ww9/wcGGf+mY7v2J4HzptnP5cpILHgSkiRp/Vi2QEKSpPWmcaDkC4qBhCRJDYricK1caWOt8F4bkiSpNzMSkiQ1WsnBlmuFgYQkSQ0KOGwgMcLShiRJ6s2MhCRJjSxtjDKQkCSpQYFXbYxhaUOSJPVmRkKSpEZORzXKQEKSpAZFedXGGJY2JElSb2YkJElqUXDYhMQIAwlJkhoMbiOu+SxtSJKk3sxISJLUJBwmq92JmWMgIUlSgwKOOEZihKUNSZLUmxkJSZIaWdoYZSAhSVKDwW3EDSTms7QhSZJ6MyMhSVKjI2VGYj4DCUmSGljaGM/ShiRJ6s2MhCRJDYpw2L+/RxhISJLUyDESowwkJElq4BiJ8czRSJKk3sxISJLUJBwu//6ez0BCkqQGBRwxkT/CT0SSJPVmRkKSpEYOthxlICFJUoMqx0iM4yciSZJ6MyMhSVKjI5Y2RhhISJLUYDAhlYn8+fxEJElSb70DiSSnJvlqkvuS7Enynq79A0keTXJXt1w0ve5KkrRaBoMtp7GsJ5OczSHgN6vqp4HXA1ckOavb9rGqOrtbdk3cS0mSVtnchFTTWFokuSDJ/Un2Jtk+ZnuSfLzbfneSc7r2M4f+mL8rydNJ3tttm/of+73HSFTVfmB/9/wHSe4DTp60Q5IkvdAl2QB8Ajgf2AfcnmRnVd07tNuFwOZueR1wNfC6qrofOHvodR4Fvjh03Meq6g+n1dep5FeSnA68FvhG13RlFx1dm+S4BY7ZlmR3kt0HDx6cRjckSVpWhytTWRqcC+ytqoeq6jngBuDieftcDHymBm4Djk2yad4+5wEPVtV3Jz33hUwcSCR5KXAT8N6qeppBRPRqBtHQfuAj446rqh1VtaWqtmzcuHHSbkiStKyKcJgfm8oCnDD3x3S3bJv3dicDjwyt72M069+yz1bg+nltS/6xfzQmCiSSvJhBEPHZqvoCQFUdqKrDVXUE+CSDqEqSJP3IE3N/THfLjnnbx6Ut6mj2SXIM8FbgT4e2N/2xfzR6j5FIEuBTwH1V9dGh9k3d+AmAtwP3TNZFSZJmw5GVu+JiH3Dq0PopwGNHuc+FwJ1VdWCuYfh5kk8CfzFpRyeZkOoNwDuBbye5q2v7HeCSJGcziIoeBn59gveQJGkmrPCEVLcDm5OcwWCw5FbgV+bts5NBmeIGBoMtvz/0hzzAJcwrayzHH/uTXLXxN4xPq3i5pyRJE6iqQ0muBG4GNgDXVtWeJJd3269h8P/bi4C9wDPAu+aOT/LjDK74mP/H/Ien/ce+U2RLktSgaL7iYjrvN5iHade8tmuGnhdwxQLHPgP85Jj2d065mwYSkiS1ap1M6oXET0SSJPVmRkKSpAZVrLv7ZEyDgYQkSU3CkbHXGLywGVpJkqTezEhIktSgsLQxjoGEJEmNVnBCqjXDT0SSJPVmRkKSpAZFOLKCE1KtFQYSkiQ1srQxyk9EkiT1ZkZCkqQGxYreRnzNMJCQJKlJOOyEVCMMrSRJUm9mJCRJamBpYzwDCUmSGlnaGGVoJUmSejMjIUlSg6pY2hjDQEKSpEbetGuUn4gkSerNjIQkSQ0KOOJgyxEGEpIkNYmljTH8RCRJUm9mJCRJajCYkMrSxnwGEpIkNfI24qP8RCRJUm9mJCRJalDE0sYYBhKSJDU6YiJ/hJ+IJEnqzYyEJEkNquCwpY0RBhKSJDVyjMQoSxuSJKk3MxKSJDUYXLXh39/zGUhIktTosDftGmEgIUlSA6fIHs8cjSRJ6m2ijESSh4EfAIeBQ1W1JcnxwOeB04GHgV+uqn+crJuSJK02x0iMM41P5Oer6uyq2tKtbwdurarNwK3duiRJa94RMpVlPVmO0Opi4Lru+XXA25bhPSRJWteSXJDk/iR7k4z8UZ6Bj3fb705yztC2h5N8O8ldSXYPtR+f5JYkD3SPx03az0kDiQK+nOSOJNu6tpOqaj9A93jiuAOTbEuyO8nugwcPTtgNSZKW19zMltNYlpJkA/AJ4ELgLOCSJGfN2+1CYHO3bAOunrd9fsUAlqFqMGkg8YaqOofByVyR5E2tB1bVjqraUlVbNm7cOGE3JElafkfqx6ayNDgX2FtVD1XVc8ANDDL+wy4GPlMDtwHHJtm0xOtOvWowUSBRVY91j48DX2Rw4gfmTqR7fHzSTkqStM6cMJeV75Zt87afDDwytL6va2vdZ1zFABqrBkej91UbSX4C+LGq+kH3/BeA3wN2ApcCV3WPX5q0k5IkrbbBzJZTGyj5xLySw3zj3qiOYp83VNVjSU4Ebknynar6Wp+OLmWSyz9PAr6YZO51PldVf5XkduDGJJcB3wPeMXk3JUlafSt4xcU+4NSh9VOAx1r3Ga4YJJmrGHyNrmpQVfunVTXoHUhU1UPAfxjT/iRw3iSdkiTpBe52YHOSM4BHga3Ar8zbZydwZZIbgNcB3+8ChIUqBnPHTLVq4BTZkiQ1WMkpsqvqUJIrgZuBDcC1VbUnyeXd9muAXcBFwF7gGeBd3eFjKwbdtquYctXAQEKSpEYrObNlVe1iECwMt10z9LyAK8YcN7Zi0G2betXAuT4lSVJvZiQkSWpRU71qY90wkJAkqUGxoldtrBmWNiRJUm9mJCRJamRpY5SBhCRJDVby8s+1xNKGJEnqzYyEJEmNzEiMMpCQJKnBlG/atW5Y2pAkSb2ZkZAkqZHzSIwykJAkqUU5RmIcSxuSJKk3MxKSJDVwHonxDCQkSWpkIDHK0oYkSerNjIQkSQ2cR2I8AwlJkhqVgcQISxuSJKk3MxKSJDVyQqpRBhKSJDUoJ6Qay9KGJEnqzYyEJEmNHGw5ykBCkqQmXv45jqUNSZLUmxkJSZIaWdoYZSAhSVIDb9o1nqUNSZLUmxkJSZJa1GAuCT2fgYQkSY2c2XKUpQ1JktSbGQlJkhoUXrUxjoGEJElNnJBqHEsbkiSpNzMSkiQ18qqNUb0DiSRnAp8fanoV8LvAscD/AA527b9TVbv6vo8kSbPCMRKjegcSVXU/cDZAkg3Ao8AXgXcBH6uqP5xGByVJ0uyaVmnjPODBqvpuYrQmSVp/qsxIjDOtwZZbgeuH1q9McneSa5McN+6AJNuS7E6y++DBg+N2kSRpphypTGVZTyYOJJIcA7wV+NOu6Wrg1QzKHvuBj4w7rqp2VNWWqtqycePGSbshSdK6kuSCJPcn2Ztk+5jtSfLxbvvdSc7p2k9N8tUk9yXZk+Q9Q8d8IMmjSe7qlosm7ec0ShsXAndW1QGAuUeAJJ8E/mIK7yFJ0qpbqas2urGHnwDOB/YBtyfZWVX3Du12IbC5W17H4A/51wGHgN+sqjuTvAy4I8ktQ8dOdRzjNEoblzBU1kiyaWjb24F7pvAekiStuqpMZWlwLrC3qh6qqueAG4CL5+1zMfCZGrgNODbJpqraX1V3DvpbPwDuA06e3qfwfBMFEkl+nEG09IWh5g8n+XaSu4GfB/7XJO8hSdIsKKYTRHSBxAlz4wS7Zdu8tzsZeGRofR+jwcCS+yQ5HXgt8I2h5iXHMR6NiUobVfUM8JPz2t45UY8kSVr/nqiqLYtsH5e2mF9YWXSfJC8FbgLeW1VPd81XAx/s9vsgg3GM727t9DjObClJUqMVnNhyH3Dq0PopwGOt+yR5MYMg4rNV9cOqwXKMY/ReG5IktagVHSNxO7A5yRnd1ZFbgZ3z9tkJ/Gp39cbrge9X1f4MJnT6FHBfVX10+IDlGMdoRkKSpBlTVYeSXAncDGwArq2qPUku77ZfA+wCLgL2As8wmFka4A3AO4FvJ7mra5u7XcWHk5zNILnyMPDrk/bVQEKSpFYrWNvo/se/a17bNUPPC7hizHF/w/jxE8syjtFAQpKkRk6RPcoxEpIkqTczEpIkNVqpmS3XEgMJSZIaFJY2xrG0IUmSejMjIUlSiwLMSIwwkJAkqZFjJEZZ2pAkSb2ZkZAkqZUZiREGEpIkNWm+T8YLiqUNSZLUmxkJSZJaWdoYYSAhSVKLckKqcSxtSJKk3sxISJLUytLGCAMJSZKaWdqYz9KGJEnqzYyEJEmtLG2MMJCQJKmVgcQISxuSJKk3MxKSJLXwNuJjGUhIktTI24iPsrQhSZJ6MyMhSVIrMxIjDCQkSWrlGIkRljYkSVJvZiQkSWoUSxsjDCQkSWpROEZiDEsbkiSpNzMSkiQ1iYMtxzCQkCSplaWNEZY2JElSb2YkJElqZUZixJIZiSTXJnk8yT1DbccnuSXJA93jcUPb3pdkb5L7k7xluTouSdKKqykt60hLaePTwAXz2rYDt1bVZuDWbp0kZwFbgdd0x/xxkg1T660kSZopSwYSVfU14Kl5zRcD13XPrwPeNtR+Q1U9W1V/D+wFzp1OVyVJWkVztxGfxrKO9B1seVJV7QfoHk/s2k8GHhnab1/XNiLJtiS7k+w+ePBgz25IkrRyUtNZ1pNpX7UxLswa+5FV1Y6q2lJVWzZu3DjlbkiSpJXQN5A4kGQTQPf4eNe+Dzh1aL9TgMf6d0+SpBmygoMtk1zQXbiwN8n2MduT5OPd9ruTnLPUsYtdLNFX30BiJ3Bp9/xS4EtD7VuTvCTJGcBm4JuTdVGSpBeW7kKFTwAXAmcBl3QXNAy7kMH/ZzcD24CrG44de7HEJFou/7we+DpwZpJ9SS4DrgLOT/IAcH63TlXtAW4E7gX+Criiqg5P2klJkl5gzgX2VtVDVfUccAODCxqGXQx8pgZuA47tqgSLHbvQxRK9LTkhVVVdssCm8xbY/0PAhybplCRJs2gFB0qOu3jhdQ37nLzEsc+7WCLJiUzImS0lSWo1vUs3T0iye2h9R1XtGFpvuXhhoX2aL3yYBgMJSZJW3hNVtWWR7S0XLyy0zzGLHHsgyaYuGzF8sURv3rRLkqQW07pioy03cDuwOckZSY5hMGv0znn77AR+tbt64/XA97uyxWLHLnSxRG9mJCRJarVCYySq6lCSK4GbgQ3AtVW1J8nl3fZrgF3ARQxmkX4GeNdix3YvfRVwY3fhxPeAd0zaVwMJSZIareSslFW1i0GwMNx2zdDzAq5oPbZrf5IFLpboy9KGJEnqzYyEJEmt1tl9MqbBQEKSpFYGEiMsbUiSpN7MSEiS1GA93gJ8GgwkJElqNb2ZLdcNSxuSJKk3MxKSJLWytDHCQEKSpEaOkRhlaUOSJPVmRkKSpFZmJEYYSEiS1MLLP8eytCFJknozIyFJUiszEiMMJCRJamUgMcLShiRJ6s2MhCRJjRxsOcqMhCRJ6s1AQpIk9WZpQ5KkVpY2RhhISJLUwgmpxrK0IUmSejMjIUlSKzMSIwwkJElqZSAxwtKGJEnqzYyEJEkNgoMtxzGQkCSplYHECEsbkiSpNzMSkiS1cB6JsQwkJElqZSAxwtKGJEnqzYyEJEmtzEiMWDIjkeTaJI8nuWeo7Q+SfCfJ3Um+mOTYrv30JP+U5K5uuWYZ+y5J0opKTWdZT1pKG58GLpjXdgvwM1X174G/A943tO3Bqjq7Wy6fTjclSdIsWjKQqKqvAU/Na/tyVR3qVm8DTlmGvkmSNFtqSss6Mo3Blu8G/nJo/Ywkf5vkr5O8caGDkmxLsjvJ7oMHD06hG5IkLaNpBREGEj+S5P3AIeCzXdN+4LSqei3wG8Dnkrx83LFVtaOqtlTVlo0bN07SDUmStEp6X7WR5FLgF4HzqqoAqupZ4Nnu+R1JHgR+Ctg9hb5KkrSq1ttAyWnoFUgkuQD4beA/VdUzQ+0bgaeq6nCSVwGbgYem0lNJklabgcSIlss/rwe+DpyZZF+Sy4A/Al4G3DLvMs83AXcn+RbwZ8DlVfXU2BeWJElHLcnxSW5J8kD3eNwC+12Q5P4ke5NsH2qf6hQOS2YkquqSMc2fWmDfm4CbWt5YkqS1ZkZKG9uBW6vqqi5A2M6gSvBDSTYAnwDOB/YBtyfZWVX3MpjC4X1VdSjJ7zOYwmHu+Aer6uyj6YxTZEuS1Go2rtq4GLiue34d8LYx+5wL7K2qh6rqOeCG7ripT+FgICFJ0so7YW4KhG7ZdhTHnlRV+wG6xxPH7HMy8MjQ+r6ubb5eUzgM814bkiS1mO4cEE9U1ZaFNib5CvCKMZve3/j6GdP2vN4vMoXDk0l+FvjzJK+pqqcXeyMDCUmSGoTx/3deDlX15gX7kRxIsqmq9ifZBDw+Zrd9wKlD66cAjw29xtSmcLC0IUnS2rITuLR7finwpTH73A5sTnJGkmOArd1xw1M4vHX+FA7dIE2OZgoHAwlJklrNxmDLq4DzkzzA4KqMqwCSvDLJLoBuMOWVwM3AfcCNVbWnO36qUzhY2pAkqdEsXP5ZVU8C541pfwy4aGh9F7BrzH7/doHX7TWFgxkJSZLUmxkJSZJazUBGYtYYSEiS1MpAYoSlDUmS1JsZCUmSWtRsDLacNQYSkiS1MpAYYSAhSVIjMxKjHCMhSZJ6MyMhSVIrMxIjDCQkSWpkaWOUpQ1JktSbGQlJklpM54Zb646BhCRJrQwkRljakCRJvZmRkCSpQXCw5TgGEpIktTKQGGFpQ5Ik9WZGQpKkRilTEvMZSEiS1MLLP8eytCFJknozIyFJUiOv2hhlICFJUisDiRGWNiRJUm9mJCRJamRpY5SBhCRJrQwkRljakCRJvZmRkCSpRVnaGMdAQpKkVgYSIyxtSJKk3sxISJLUwNuIj7dkRiLJtUkeT3LPUNsHkjya5K5uuWho2/uS7E1yf5K3LFfHJUlacVXTWdaRltLGp4ELxrR/rKrO7pZdAEnOArYCr+mO+eMkG6bVWUmSNFuWDCSq6mvAU42vdzFwQ1U9W1V/D+wFzp2gf5IkzYzUdJb1ZJLBllcmubsrfRzXtZ0MPDK0z76ubUSSbUl2J9l98ODBCbohSdIKqCku60jfQOJq4NXA2cB+4CNde8bsO/Yjq6odVbWlqrZs3LixZzckSdJq6nXVRlUdmHue5JPAX3Sr+4BTh3Y9BXisd+8kSZohObLaPZg9vTISSTYNrb4dmLuiYyewNclLkpwBbAa+OVkXJUmaEZY2RiyZkUhyPfBzwAlJ9gH/G/i5JGcz+DgeBn4doKr2JLkRuBc4BFxRVYeXpeeSJGnVLRlIVNUlY5o/tcj+HwI+NEmnJEmaRevtiotpcIpsSZJaFDMxIVWS45PckuSB7vG4Bfa7oJsccm+S7UPtU51U0kBCkqS1ZTtwa1VtBm7t1p+nmwzyE8CFwFnAJd2kkXOmNqmkgYQkSY1mZEKqi4HruufXAW8bs8+5wN6qeqiqngNu6I5b6nWPelJJAwlJklpN76qNE+YmZeyWbUfRi5Oqaj9A93jimH2WmiByokklh3n3T0mSVt4TVbVloY1JvgK8Ysym9ze+/mITRF4NfLBb/yCDSSXfvcQxCzKQkCSpwUreRryq3rxgP5IDSTZV1f5uXqfHx+y24ASR055U0tKGJEktpnXFxuS3Ed8JXNo9vxT40ph9bgc2JzkjyTEMBlHuhOlPKmlGQpKkteUq4MYklwHfA94BkOSVwJ9U1UVVdSjJlcDNwAbg2qra0x3/4WlOKmkgIUlSo1mYkKqqngTOG9P+GHDR0PouYNeY/d65yGsf9aSSBhKSJLWagUBi1jhGQpIk9WZGQpKkRrNQ2pg1BhKSJLUo4IiRxHyWNiRJUm9mJCRJamVCYoSBhCRJjRwjMcrShiRJ6s2MhCRJrSaf3nrdMZCQJKmRpY1RljYkSVJvZiQkSWpReNXGGAYSkiQ1CBDHSIwwkJAkqdWR1e7A7HGMhCRJ6s2MhCRJjSxtjDKQkCSphYMtx7K0IUmSejMjIUlSk3JmyzEMJCRJauTMlqMsbUiSpN7MSEiS1MrSxggDCUmSWhTECalGWNqQJEm9mZGQJKmVpY0RBhKSJLUyjhhhaUOSJPVmRkKSpEbea2PUkoFEkmuBXwQer6qf6do+D5zZ7XIs8H+r6uwkpwP3Afd3226rqsun3WlJklaFgcSIlozEp4E/Aj4z11BV/23ueZKPAN8f2v/Bqjp7Sv2TJEkzbMlAoqq+1mUaRiQJ8MvAf55yvyRJmi0FOI/EiEkHW74ROFBVDwy1nZHkb5P8dZI3LnRgkm1JdifZffDgwQm7IUnS8gpFajrLejJpIHEJcP3Q+n7gtKp6LfAbwOeSvHzcgVW1o6q2VNWWjRs3TtgNSZK0GnpftZHkRcAvAT8711ZVzwLPds/vSPIg8FPA7gn7KUnS6ltn2YRpmOTyzzcD36mqfXMNSTYCT1XV4SSvAjYDD03YR0mSZoOBxIglSxtJrge+DpyZZF+Sy7pNW3l+WQPgTcDdSb4F/BlweVU9Nc0OS5Kk2dFy1cYlC7T/2pi2m4CbJu+WJEkzxqs2xnJmS0mSGq23Ky6mwXttSJK0hiQ5PsktSR7oHo9bYL8LktyfZG+S7UPtn09yV7c8nOSurv30JP80tO2alv6YkZAkqdVsZCS2A7dW1VVdgLAd+O3hHZJsAD4BnA/sA25PsrOq7p327NRmJCRJalKDQGIay2QuBq7rnl8HvG3MPucCe6vqoap6DrihO+6Hhmannn/hxFExkJAkaeWdMDe7c7dsO4pjT6qq/QDd44lj9jkZeGRofV/XNqz37NTDLG1IktSimGZp44mq2rLQxiRfAV4xZtP7G18/Y9rmd36h2amfTPKzwJ8neU1VPb3YGxlISJLUaoUu/6yqNy+0LcmBJJuqan+STcDjY3bbB5w6tH4K8NjQa0xtdmpLG5IkrS07gUu755cCXxqzz+3A5iRnJDmGwSSSO4e2j52duhukydHMTm1GQpKkRjMyj8RVwI3dTNPfA94BkOSVwJ9U1UVVdSjJlcDNwAbg2qraM/QaC81O/XtJDgGHaZyd2kBCkqRWMxBIVNWTwHlj2h8DLhpa3wXsWuA1fm1MW6/ZqS1tSJKk3sxISJLUooAjq5+RmDUGEpIkNZnKZFLrjqUNSZLUmxkJSZJamZEYYSAhSVIrA4kRljYkSVJvZiQkSWrhVRtjGUhIktSkoFboZhtriKUNSZLUmxkJSZJaOdhyhIGEJEktHCMxlqUNSZLUmxkJSZJaWdoYYSAhSVIrA4kRljYkSVJvZiQkSWri3T/HMZCQJKlFAUeckGo+SxuSJKk3MxKSJLWytDHCQEKSpFYGEiMMJCRJalLObDmGYyQkSVJvZiQkSWpRUN5GfISBhCRJrSxtjLC0IUmSejMjIUlSK6/aGLFkRiLJqUm+muS+JHuSvKdrPz7JLUke6B6PGzrmfUn2Jrk/yVuW8wQkSVoRVYOZLaexrCMtpY1DwG9W1U8DrweuSHIWsB24tao2A7d263TbtgKvAS4A/jjJhuXovCRJWl1LBhJVtb+q7uye/wC4DzgZuBi4rtvtOuBt3fOLgRuq6tmq+ntgL3DulPstSdLKq5rOso4c1RiJJKcDrwW+AZxUVfthEGwkObHb7WTgtqHD9nVt819rG7AN4LTTTjvqjkuStNJqnZUlpqH5qo0kLwVuAt5bVU8vtuuYtpHwq6p2VNWWqtqycePG1m5IkqQZ0pSRSPJiBkHEZ6vqC13zgSSbumzEJuDxrn0fcOrQ4acAj02rw5IkrY71V5aYhparNgJ8Crivqj46tGkncGn3/FLgS0PtW5O8JMkZwGbgm9PrsiRJq6AYTEg1jWUdaclIvAF4J/DtJHd1bb8DXAXcmOQy4HvAOwCqak+SG4F7GVzxcUVVHZ52xyVJ0upbMpCoqr9h/LgHgPMWOOZDwIcm6JckSbPHe22McGZLSZIaFFDrrCwxDd5rQ5Ik9WZGQpKkFlWWNsYwIyFJUqM6UlNZJrHYva7m7XdtkseT3NN6fJ97ZRlISJK0toy919UYn2Zwz6um4/veK8tAQpKkVnVkOstkFrrX1fO7WvU14KmjOL7XvbJmYozEHXfc8USS/wc8sdp9mdAJrO1zWOv9h7V/Dmu9/7D2z2Gt9x/W/jn06f+/WY6ODPsB/3jzV+rPTpjSy/2rJLuH1ndU1Y7GYxe611Wrie6VNd9MBBJVtTHJ7qrastp9mcRaP4e13n9Y++ew1vsPa/8c1nr/Ye2fw6z2v6rGlQmWRZKvAK8Ys+n9y/m2Y9qWHNAxE4GEJEn6kap680Lbkix0r6tWU71XlmMkJElaWxa619Wkx/e6V9YsBRKttaFZttbPYa33H9b+Oaz1/sPaP4e13n9Y++ew1vu/3K4Czk/yAHB+t06SVybZNbdTkuuBrwNnJtnX3RtrweOrag8wd6+sv6LxXlkpb4kqSZJ6mqWMhCRJWmMMJCRJUm8zEUgkuaCbjnNvkoVm6JoZSU5N8tUk9yXZk+Q9XfsHkjya5K5uuWi1+7qYJA8n+XbX191dW9PUq6styZlDn/NdSZ5O8t5Z/w7GTVk77elql9MC/f+DJN9JcneSLyY5tms/Pck/DX0X16xax4cscA4L/tzM2ncAC57D54f6/3CSu7r2mfseFvk3dM38LmhIVa3qAmwAHgReBRwDfAs4a7X7tUSfNwHndM9fBvwdcBbwAeC3Vrt/R3EeDwMnzGv7MLC9e74d+P3V7mfjz9A/MJiQZqa/A+BNwDnAPUt95t3P1LeAlwBndL8nG2aw/78AvKh7/vtD/T99eL9ZWRY4h7E/N7P4HSx0DvO2fwT43Vn9Hhb5N3TN/C64/GiZhYzEucDeqnqoqp4DbmAwTefMqqr9VXVn9/wHwH00zP61RjRNvTpjzgMerKrvrnZHllLjp6yd6nS1y2lc/6vqy1V1qFu9jcG15zNrge9gITP3HcDi55AkwC8D169op47CIv+GrpnfBf3ILAQSJwOPDK03Tck5K5KcDrwW+EbXdGWX4r12VssCQwr4cpI7kmzr2p43dSpwtFOvroatPP8fzbX0HcDCn/la/N14N/CXQ+tnJPnbJH+d5I2r1alG435u1uJ38EbgQFU9MNQ2s9/DvH9D19PvwgvGLAQSvabknAVJXgrcBLy3qp4GrgZeDZwN7GeQXpxlb6iqc4ALgSuSvGm1O3S0khwDvBX4065prX0Hi1lTvxtJ3g8cAj7bNe0HTquq1wK/AXwuyctXq39LWOjnZk19B51LeH5gPbPfw5h/QxfcdUzbrH8PLxizEEj0mpJztSV5MYNfgM9W1RcAqupAVR2uqiPAJ5nx1FtVPdY9Pg58kUF/D2QwZSrpN/XqSrsQuLOqDsDa+w46C33ma+Z3I8mlwC8C/71qUNTu0tBPds/vYFDX/qnV6+XCFvm5WTPfAUCSFwG/BHx+rm1Wv4dx/4ayDn4XXohmIZC4Hdic5Izur8utDKbpnFldDfJTwH1V9dGh9k1Du70duGf+sbMiyU8kednccwYD5u5h8qlXV9rz/vpaS9/BkKlOV7vSklwA/Dbw1qp6Zqh9Y5IN3fNXMej/Q6vTy8Ut8nOzJr6DIW8GvlNV++YaZvF7WOjfUNb478IL1mqP9uz+eLmIwajdB4H3r3Z/Gvr7Hxmk1e4G7uqWi4D/A3y7a98JbFrtvi5yDq9iMAr6W8Ceuc8d+EngVuCB7vH41e7rIufw48CTwL8eapvp74BB0LMf+BcGf2VdtthnzuBOfw8C9wMXzmj/9zKoX8/9LlzT7ftfu5+tbwF3Av9ltfu/yDks+HMza9/BQufQtX8auHzevjP3PSzyb+ia+V1w+dHiFNmSJKm3WShtSJKkNcpAQpIk9WYgIUmSejOQkCRJvRlISJKk3gwkJElSbwYSkiSpt/8PHj3Y4dqeXo8AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 648x504 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig,ax = plt.subplots(figsize=(9,7))\n",
    "h1 = ax.imshow(hd, aspect='auto')\n",
    "fig.colorbar(h1, ax=ax)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "environmental-wallace",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-08T04:41:45.470629Z",
     "start_time": "2022-09-08T04:41:45.124883Z"
    }
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Input contains NaN, infinity or a value too large for dtype('float64').",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-6980d89c75ed>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mclusters\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mSpectralClustering\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_clusters\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mn_embryos\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0massign_labels\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"discretize\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhd\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mclusters\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlabels_\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/mnt/geofflab/jupyterhub/anaconda3/lib/python3.7/site-packages/sklearn/cluster/_spectral.py\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, X, y)\u001b[0m\n\u001b[1;32m    478\u001b[0m         \"\"\"\n\u001b[1;32m    479\u001b[0m         X = self._validate_data(X, accept_sparse=['csr', 'csc', 'coo'],\n\u001b[0;32m--> 480\u001b[0;31m                                 dtype=np.float64, ensure_min_samples=2)\n\u001b[0m\u001b[1;32m    481\u001b[0m         allow_squared = self.affinity in [\"precomputed\",\n\u001b[1;32m    482\u001b[0m                                           \"precomputed_nearest_neighbors\"]\n",
      "\u001b[0;32m/mnt/geofflab/jupyterhub/anaconda3/lib/python3.7/site-packages/sklearn/base.py\u001b[0m in \u001b[0;36m_validate_data\u001b[0;34m(self, X, y, reset, validate_separately, **check_params)\u001b[0m\n\u001b[1;32m    418\u001b[0m                     \u001b[0;34mf\"requires y to be passed, but the target y is None.\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    419\u001b[0m                 )\n\u001b[0;32m--> 420\u001b[0;31m             \u001b[0mX\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcheck_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mX\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mcheck_params\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    421\u001b[0m             \u001b[0mout\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mX\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    422\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/mnt/geofflab/jupyterhub/anaconda3/lib/python3.7/site-packages/sklearn/utils/validation.py\u001b[0m in \u001b[0;36minner_f\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     70\u001b[0m                           FutureWarning)\n\u001b[1;32m     71\u001b[0m         \u001b[0mkwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0marg\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marg\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparameters\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 72\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     73\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0minner_f\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     74\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/mnt/geofflab/jupyterhub/anaconda3/lib/python3.7/site-packages/sklearn/utils/validation.py\u001b[0m in \u001b[0;36mcheck_array\u001b[0;34m(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, estimator)\u001b[0m\n\u001b[1;32m    643\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mforce_all_finite\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    644\u001b[0m             _assert_all_finite(array,\n\u001b[0;32m--> 645\u001b[0;31m                                allow_nan=force_all_finite == 'allow-nan')\n\u001b[0m\u001b[1;32m    646\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    647\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mensure_min_samples\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/mnt/geofflab/jupyterhub/anaconda3/lib/python3.7/site-packages/sklearn/utils/validation.py\u001b[0m in \u001b[0;36m_assert_all_finite\u001b[0;34m(X, allow_nan, msg_dtype)\u001b[0m\n\u001b[1;32m     97\u001b[0m                     \u001b[0mmsg_err\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     98\u001b[0m                     (type_err,\n\u001b[0;32m---> 99\u001b[0;31m                      msg_dtype if msg_dtype is not None else X.dtype)\n\u001b[0m\u001b[1;32m    100\u001b[0m             )\n\u001b[1;32m    101\u001b[0m     \u001b[0;31m# for object dtype data, we only check for NaNs (GH-13254)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Input contains NaN, infinity or a value too large for dtype('float64')."
     ]
    }
   ],
   "source": [
    "clusters = SpectralClustering(n_clusters=n_embryos,assign_labels=\"discretize\").fit(hd)\n",
    "print(clusters.labels_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "equipped-tribute",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-09-08T04:41:45.472760Z",
     "start_time": "2022-09-08T04:37:54.974Z"
    }
   },
   "outputs": [],
   "source": [
    "print(adjusted_rand_score(true_clusters,clusters.labels_))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "equal-cliff",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
